lib/Importer/ProtobufLoader.cpp

```diff
@@ -73,12 +73,50 @@ ProtobufLoader::getStaticPlaceholderByNameOrNull(llvm::StringRef name) const {
   return (res && res->isStatic()) ? res : nullptr;
 }
 
+bool ProtobufLoader::isOpRequestedInFP16Precision(llvm::StringRef name) const {
+  auto opPrecisionUpdateItr =
+      std::find(operatorPrecisionUpdateInputOutputNames_.begin(),
+                operatorPrecisionUpdateInputOutputNames_.end(), name);
+  return opPrecisionUpdateItr != operatorPrecisionUpdateInputOutputNames_.end();
+}
+
+void ProtobufLoader::updateNodeOutputPrecision(llvm::StringRef outputName,
+                                               NodeValue NV) {
+  // If output name is specified to run in FP16 precision and output type
+  // is fp16 add convert to fp32 node and update nodeValueByName_ map,
+  // so that next operator in the model receives appropriate nodeValue.
+  if (isOpRequestedInFP16Precision(outputName) &&
+      NV.getElementType() == ElemKind::Float16Ty) {
+    std::string opName = outputName.str() + "_convToFP32";
+    auto *CT = G_->createConvertTo(opName, NV, ElemKind::FloatTy);
+    nodeValueByName_[outputName] = CT->getResult();
+  } else {
+    nodeValueByName_[outputName] = NV;
+  }
+}
+
 Constant *ProtobufLoader::getConstantByNameOrNull(llvm::StringRef name) const {
   auto it = nodeValueByName_.find(name);
   if (it == nodeValueByName_.end()) {
     return nullptr;
   }
-  auto *res = llvm::dyn_cast<Constant>(it->second.getNode());
+  NodeValue NV = it->second;
+  auto *res = llvm::dyn_cast<Constant>(NV.getNode());
+  // If name is specified to run in FP16 precision and output type
+  // is FP32 add ConvertTo FP16 node to the original nodeValue and return the
+  // updated nodeValue.
+  // For Constant inputs, convertTofp16 nodes cannot be added
+  // because it won't be Constant anymore and cannot access the tensor
+  // payload. So for Constant inputs change tensor datatype and tensor
+  // payload to fp16.
```

Comment:
So to recap the situation:
1. The Constant is specified based on its name in the yaml to be FP16
2. You load the Constant originally in FP32, usually by calling `createAndRegisterConstant()`
3. You leave it as FP32 until the first time it's accessed, at which point you convert it to FP16

Would it make more sense to just do this conversion when the Constant is initially loaded, e.g. inside `createAndRegisterConstant()`?

Reply:
I am converting the constant to FP16 only when it is passed as input to operator required in fp16. In createAndRegisterConstant() method when Constant is initially loaded, currently we don't have information about the operators which uses the loaded constant as input.  I do not see the advantages in making the changes to convert the constants in createAndRegisterConstant()

Reply:
Ok so my point 1 was incorrect -- the yaml only specifies the ops that should be in fp16, not the Constants themselves, so determining what should be fp16 makes sense to do lazily only based on the Constants users.

One issue with this approach then is if there are two nodes that use a Constant, and only one of them wants it in FP16. This seems like something we should throw an error on right? E.g. adding something below before we `return res;`:

```
if (res && res->getOutput().getElementType() == ElemKind::Float16Ty) {
  CHECK(isOpRequestedInFP16Precision(name)) << "Expected FP32 Constant but found FP16 for " << name;
}
```

WDYT?

Reply:
This error check throws errors for models(Caffe2ImporterTest, OnnxExporterTest) with constants which are already in FP16(e.x FbFCPacked, fp16FullyConnected in onnx). I am not sure whether a constant node can have more than one user, so this check seems like an overkill.